### ISSUE TITLE
Cross Attention in training and generation

### ISSUE THREAD

**Issue by athena913** [tencent-ailab/IP-Adapter#378]
Hi,
Thank you for releasing your  code. I would like to understand where is the decoupled cross-attention being used in the code,  as stated in the paper. In the code, I only say concatenation. I would appreciate any explanation you can provide - thank you.

1) In the training code in tutorial_train.py in the IPAdapter class, the encoder state is being concatenated with the output of the image projection model. But in the architectural diagram of the paper, it is being input through a decoupled  cross-attention model, which I don't see in the code below.
 ```
 def forward(self, noisy_latents, timesteps, encoder_hidden_states, image_embeds):
        ip_tokens = self.image_proj_model(image_embeds)
        encoder_hidden_states = torch.cat([encoder_hidden_states, ip_tokens], dim=1)
        # Predict the noise residual
        noise_pred = self.unet(noisy_latents, timesteps, encoder_hidden_states).sample
        return noise_pred
```

2) Likewise, in the generator code, the optional text prompt and image prompt embeddings (using the trained projection model) are concatenated and input to the stable diffusion model for generation. I don't see any cross attention during generation.

 `prompt_embeds = torch.cat([prompt_embeds_, image_prompt_embeds], dim=1)`

I dont see the following ip_layers being used during generation.

```
 ip_layers = torch.nn.ModuleList(self.pipe.unet.attn_processors.values())
 ip_layers.load_state_dict(state_dict["ip_adapter"])

```

**Comment by xiaohu2015** [comment]
see https://github.com/tencent-ailab/IP-Adapter/blob/main/ip_adapter/attention_processor.py#L82

**Comment by athena913** [comment]
Thank you. I understand where your have implemented the cross-attention. But is the cross-attention being used  during generation? It looks like the image is embedded using the projection layer and then concatenated with the (optional) text embedding for without using cross-attention during generation . Is this correct?

**Comment by xiaohu2015** [comment]
@athena913  it is used, see https://github.com/tencent-ailab/IP-Adapter/blob/main/ip_adapter/ip_adapter.py#L94

**Comment by nguyendung1612** [comment]
Thanks for your contribution but can you explain in more detail?

1) In tutorial_train.py, I see 'adapter_modules' is just initialized but not used during training. How can it update the parameter?
![image](https://github.com/user-attachments/assets/35963dcd-eee6-4028-bf95-1d4875cfa52c)

2) > @athena913 it is used, see https://github.com/tencent-ailab/IP-Adapter/blob/main/ip_adapter/ip_adapter.py#L94

In set_ip_adapter func, you declare a new 'unet' and use it to call 'set_attn_processor'. Does this have any effect on 'unet' of pipeline?
![image](https://github.com/user-attachments/assets/35b34658-6f21-4eef-bb99-6869793711bd)

3) In load_ip_adapter func, You initialize 'ip_layers' to load weight of 'ip_adapter'. but I don't see it used anywhere in the source code.
![image](https://github.com/user-attachments/assets/97eb5e4e-a702-4c9a-aa2f-73d13df2da87)
